Code/Tools/AssetProcessor/native/AssetManager/AssetCatalog.cpp

```diff
@@ -589,6 +589,53 @@ namespace AssetProcessor
         m_catalogIsDirty = true;
     }
 
+    void AssetCatalog::OnConnect(unsigned int connectionId, ::Connection* connection)
+    {
+        if(!connection || connection->Status() != ::Connection::ConnectionStatus::Connected)
+        {
+            return;
+        }
+
+        // Send out a message for each asset to make sure the connected tools are aware of the existence of all previously built assets
```

Comment:
Seems very expensive to push notifications for every asset on connect.  What impact does this have on startup time? 

Concern also that we are not batching and time in function scales directly with number of assets.

Wouldn't it be easier to just reprocess or part process AssetCache when its not found but product assets exists and send a general update mechanism - ie identify delta and send just that or some notices to reload AssetCache after its rebuilt?

Just concerned we are overindexing on the one off failure case, at a cost of general performance.

Reply:
I did some quick timing, it seems to take about 4 seconds for the AutomatedTest project.

Doing a partial update would be fairly complicated.  It's not just the editor connecting, there are asset builders and the launcher, and potentially other applications.  Depending on when the application connects, the info that has to be sent out will be different.  Basically the delta will be what's in the assetcatalog.xml at the time the application reads it and what's in the AP's in-memory asset catalog.  But there is no way to know exactly when the other application read the assetcatalog.  The only easy case is when the assetcatalog doesn't exist yet, in which case the delta is just everything.

This could be batched, I'm not sure how much overhead time is spent doing individual messages, but that is a bit more involved fix that either requires adding a new message type or changing the existing one.